app/controllers/doorkeeper/tokens_controller.rb

```diff
@@ -113,19 +113,53 @@ def revoke_token
       # The authorization server responds with HTTP status code 200 if the token
       # has been revoked successfully or if the client submitted an invalid
       # token
-      token.revoke if token&.accessible?
+      case token_type
+      when :access_token
+        token.revoke if token.accessible?
+      when :refresh_token
+        token.revoke unless token.revoked?
+      end
     end
 
     def token
-      @token ||=
+      token_and_type&.[](:token)
+    end
+
+    def token_type
+      token_and_type&.[](:type)
+    end
+
+    def token_and_type
```

Comment:
I admit that this implementation is pretty awkward, I went with the simplest / quickest just to show the issue.  I think I would rather create an object here that is the pair instead, but not sure if y'all would think that is overkill or would match your patterns, just let me know and I can refactor this however you want it to look.

Reply:
Agree, maybe better to have some clear DTO object. Can we refactor it please? :pray:  We have some (well not good actually) examples like OAuth::Client

Reply:
I went a slightly different direction and refactored in [5dd6454](https://github.com/ransombriggs/doorkeeper/commit/5dd6454dc940aa0d50b505e4308e39d37e8ec8ba) into classes that respond to `revocable?` and `revoke`.  I think it is cleaner than this the original code, but can keep tweaking.